src/effects/effectchain.cpp

```diff
@@ -150,38 +154,61 @@ void EffectChain::setEnabled(bool enabled) {
     emit(enabledChanged(enabled));
 }
 
-void EffectChain::enableForChannel(const ChannelHandleAndGroup& handle_group) {
-    if (!m_bAddedToEngine) {
+void EffectChain::enableForInputChannel(const ChannelHandleAndGroup& handle_group) {
+    // TODO(Be): remove m_enabledChannels from this class and move this logic
+    // to EffectChainSlot
+    bool bWasAlreadyEnabled = m_enabledInputChannels.contains(handle_group);
+    if (!bWasAlreadyEnabled) {
+        m_enabledInputChannels.insert(handle_group);
+    }
+
+    // The allocation of EffectStates below may be expensive, so avoid it if
+    // not needed.
+    if (!m_bAddedToEngine || bWasAlreadyEnabled) {
         return;
     }
-    if (!m_enabledChannels.contains(handle_group)) {
-        m_enabledChannels.insert(handle_group);
 
-        EffectsRequest* request = new EffectsRequest();
-        request->type = EffectsRequest::ENABLE_EFFECT_CHAIN_FOR_CHANNEL;
-        request->pTargetChain = m_pEngineEffectChain;
-        request->channel = handle_group.handle();
-        m_pEffectsManager->writeRequest(request);
+    EffectsRequest* request = new EffectsRequest();
+    request->type = EffectsRequest::ENABLE_EFFECT_CHAIN_FOR_INPUT_CHANNEL;
+    request->pTargetChain = m_pEngineEffectChain;
+    request->channel = handle_group.handle();
+
+    // Allocate EffectStates here in the main thread to avoid allocating
+    // memory in the realtime audio callback thread.
+    auto pEffectStatesMapArray = std::make_unique<EffectStatesMapArray>();
 
-        emit(channelStatusChanged(handle_group.name(), true));
+    //TODO: get actual configuration of engine
+    const mixxx::EngineParameters bufferParameters(
+          mixxx::AudioSignal::SampleRate(96000),
+          MAX_BUFFER_LEN / mixxx::kEngineChannelCount);
+
+    for (int i = 0; i < m_effects.size(); ++i) {
+        auto pStatesMap = std::make_unique<EffectStatesMap>();
+        if (m_effects[i] != nullptr) {
+            for (const auto& outputChannel : m_pEffectsManager->registeredOutputChannels()) {
```

Comment:
We have the same loop inside EffectProcessorImpl::initialize(), can we de-duplicate is somhow? 
I think to a common function that returns EffectStatesMap (EffectStatesArray)

Reply:
By the way, a EffectStatesMap has 2 kB that is a lot of memory just to store two pointers. Any idea to reduce this?

Reply:
We could reduce ChannelHandleMap::kMaxExpectedGroups to a more reasonable value. But in the grand scheme of things I'm really not worried about 2 kB.